clients/src/test/java/org/apache/kafka/clients/producer/internals/RoundRobinPartitionerTest.java

```diff
@@ -0,0 +1,125 @@
+        // always pick an available partition,
+        // and (2) the available partitions are selected in a round robin way.
+        int countForPart0 = 0;
+        int countForPart2 = 0;
+        for (int i = 1; i <= 100; i++) {
+            int part = partitioner.partition(topic, null, null, null, null, cluster);
+            assertTrue("We should never choose a leader-less node in round robin", part == 0 || part == 2);
+            if (part == 0)
+                countForPart0++;
+            else
+                countForPart2++;
+        }
+        assertEquals("The distribution between two available partitions should be even", countForPart0, countForPart2);
+    }
+
+    @Test
+    public void testRoundRobinWithKeyBytes() throws InterruptedException {
+        final String topicA = "topicA";
+        final String topicB = "topicB";
+
+        List<PartitionInfo> allPartitions = asList(new PartitionInfo(topicA, 0, node0, nodes, nodes),
+                new PartitionInfo(topicA, 1, node1, nodes, nodes), new PartitionInfo(topicA, 2, node2, nodes, nodes),
+                new PartitionInfo(topicB, 0, node0, nodes, nodes));
+        Cluster testCluster = new Cluster("clusterId", asList(node0, node1, node2), allPartitions,
+                Collections.<String>emptySet(), Collections.<String>emptySet());
+
+        final Map<Integer, Integer> partitionCount = new HashMap<>();
+
+        for (int i = 0; i < 30; ++i) {
+            int partition = partitioner.partition(topicA, null, keyBytes, null, null, testCluster);
+            Integer count = partitionCount.get(partition);
+            if (null == count)
+                count = 0;
+            partitionCount.put(partition, count + 1);
+
+            if (i % 5 == 0) {
+                partitioner.partition(topicB, null, keyBytes, null, null, testCluster);
+            }
+        }
+
+        assertEquals(10, (int) partitionCount.get(0));
+        assertEquals(10, (int) partitionCount.get(1));
+        assertEquals(10, (int) partitionCount.get(2));
+    }
+    
+    @Test
+    public void testRoundRobinWithNullKeyBytes() throws InterruptedException {
+        final String topicA = "topicA";
+        final String topicB = "topicB";
+
+        List<PartitionInfo> allPartitions = asList(new PartitionInfo(topicA, 0, node0, nodes, nodes),
+                new PartitionInfo(topicA, 1, node1, nodes, nodes), new PartitionInfo(topicA, 2, node2, nodes, nodes),
+                new PartitionInfo(topicB, 0, node0, nodes, nodes));
+        Cluster testCluster = new Cluster("clusterId", asList(node0, node1, node2), allPartitions,
+                Collections.<String>emptySet(), Collections.<String>emptySet());
+
+        final Map<Integer, Integer> partitionCount = new HashMap<>();
+
+        for (int i = 0; i < 30; ++i) {
+            int partition = partitioner.partition(topicA, null, null, null, null, testCluster);
+            Integer count = partitionCount.get(partition);
```

Comment:
Can we use `computeIfAbsent` here?

Reply:
No we cannot use computeIfAbsent here. Not sure how this is applicable here. Also, computeIfAbsent is only meaningful for initialisation.

Reply:
ok, it's subjective. You can get the counter by 
1. `Integer count = partitionCount.computeIfAbsent(partition, x -> 0);` [or]
2. `Integer count = partitionCount.getOrDefault(partiton, 0);` to avoid null check.

Or, use Java8 merge to increment the map value directly:
3. `partitionCount.merge(partition, 1, Integer::sum);`

Reply:
The version with computeIfAbsent seems simpler to me, why not use it here?